keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,170 @@
+  - "@wojtek-t"
+editor: TBD
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Volume Scale and Performance Testing Plan
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [WIP Test Environment](#wip-test-environment)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### WIP Test Environment
+
+The tests should be developed and run in sig-scalability’s test framework and infrastructure.
+
+TODO details and links
+
+### Test Portability
+
+In order to make it easier for vendors to run these tests against their volume
+drivers, the tests for persistent volumes should only use the default
+StorageClass configured for the cluster.
+
+Therefore these tests have the following prerequisites:
+
+* A Kubernetes cluster is brought up in an appropriate environment that can
+  support the storage provider.
+* The storage backend and driver is already installed and configured.
+* A default StorageClass for that driver been installed in the cluster.
+
+### Test Cases
+
+#### Pod Startup
+
+These tests should measure how long it takes to start up a pod with volumes, assuming that the volume has already been provisioned.
+
+For each volume type:
+
+* Create many pods with 1 volume each.
+* Create 1 pod with many volumes.
+  * For PVC types, test with and without PV.NodeAffinity as one more dimension.
+  * Measure pod startup time, with a breakdown of time spent in volume scheduling, attach and mount operations.
```

Comment:
Yes, we have scalability tests today that have the infrastructure in place to measure pod startup latency.  As part of this work, we may have to enhance it if we also want to pull out finer grained metrics for volume operations.

Reply:
I guess I should have this comment here: https://github.com/kubernetes/enhancements/pull/812#discussion_r255521198

Reply:
BTW - do you already generate events on e.g. attach and mount operations? Or would they have to be added? If the latter, we should probably think more if this is the right approach (I'm not convinced - the events trick was done 3+ years ago with lack of time to do anything better).

Reply:
We have events on attach, and recently removed events for mount because they were too noisy for the remount volume types (secrets, etc).  

I was not leaning towards using events for measuring attach/mount time because we have found that events can be unreliable and dropped/gc.

Reply:
> I was not leaning towards using events for measuring attach/mount time because we have found that events can be unreliable and dropped/gc.

Yes - I'm also not a fan of using event due to reasons you mentioned. They may be used to extend the information with some optional data which e.g. helps with debugging, but if we lose it, then it's not end of the world.
So yeah - if we can find out a more robust mechanism, I'm all for it.